gpfit/fit_constraintset.py

```diff
@@ -0,0 +1,160 @@
+                else:
+                    self.constraint = (lhs == rhs)
+            else:
+                self.constraint = (lhs == rhs)
+        else:
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh >= rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = (lhs >= rhs)
+            else:
+                self.constraint = (lhs >= rhs)
+
+        self.bounds = {}
+        for i, dvar in enumerate(self.dvars):
+            self.bounds[dvar] = [fitdata["lb%d" % i],
+                                 fitdata["ub%d" % i]]
+
+        ConstraintSet.__init__(self, [self.constraint])
+
+    def get_fitdata(self):
+        " return fit data "
+        return self.fitdata
+
+    def get_dataframe(self):
+        " return a pandas DataFrame of fit parameters "
+        import pandas as pd
+        df = pd.DataFrame(self.fitdata.values()).transpose()
+        df.columns = self.fitdata.keys()
+        return df
+
+    def process_result(self, result):
+        """
+        make sure fit result is within bounds of fitted data
+        if data comes from Xfoil and airfoil is provided check against xfoil
+        """
+        super(FitCS, self).process_result(result)
+
+        if amax([abs(result["sensitivities"]["constants"][self.mfac])]) < 1e-5:
+            return None
+
+        if self.airfoil:
+            from .xfoilWrapper import xfoil_comparison
+            cl, re = 0.0, 0.0
+            for dvar in self.dvars:
+                if "Re" in str(dvar):
+                    re = result(dvar)
+                if "C_L" in str(dvar):
+                    cl = result(dvar)
+            cd = result(self.ivar)
+            if not hasattr(cl, "__len__") and hasattr(re, "__len__"):
+                cl = [cl]*len(re)
+            err, cdx = xfoil_comparison(self.airfoil, cl, re, cd)
+            ind = where(err > 0.05)[0]
+            for i in ind:
+                msg = ("Drag error for %s is %.2f. Re=%.1f; CL=%.4f;"
+                       " Xfoil cd=%.6f, GP sol cd=%.6f" %
+                       (", ".join(self.ivar.descr["models"]), err[i], re[i],
+                        cl[i], cd[i], cdx[i]))
+                print "Warning: %s" % msg
+        else:
+            for dvar in self.dvars:
+                if isinstance(dvar, NomialArray):
+                    num = [result(x) for x in dvar]
+                else:
+                    num = result(dvar)
+                direct = None
+                if any(x < self.bounds[dvar][0] for x in hstack([num])):
+                    direct = "lower"
+                    bnd = self.bounds[dvar][0]
+                if any(x > self.bounds[dvar][1] for x in hstack([num])):
+                    direct = "upper"
+                    bnd = self.bounds[dvar][1]
+
+                if direct:
+                    msg = ("Variable %.100s could cause inaccurate result"
```

Comment:
Been seeing this message when running gassolar, love it, but I think "exceeds" should be exchanged for "is above"/"is below" to make for a cleaner read the first time you see this message.